pkg/scheduler/framework/plugins/clusteraffinity/types.go

```diff
@@ -6,101 +6,298 @@ Licensed under the MIT license.
 
-// Matches returns true if the cluster matches the label selector.
-func (at *affinityTerm) Matches(cluster *clusterv1beta1.MemberCluster) bool {
-	return at.selector.Matches(labels.Set(cluster.Labels))
+// retrieveResourcePropertyValueFrom retrieves a resource property value from a member cluster.
+//
+// Note that it will return nil if the property is not available for the cluster;
+// the zero value of resource.Quantity, i.e., resource.Quantity{}, is a valid
+// quantity.
+func retrieveResourcePropertyValueFrom(cluster *clusterv1beta1.MemberCluster, name string) (*resource.Quantity, error) {
+	// Split the name into two segments, the capacity type, and the resource name.
+	//
+	// As a pre-defined rule, all the resource properties are assigned a label name of the format
+	// `[PREFIX]/[CAPACITY_TYPE]-[RESOURCE_NAME]`; for example, the allocatable CPU capacity of a
+	// a cluster has the label name, `resources.kubernetes-fleet.io/allocatable-cpu`.
+	segs := strings.Split(name, "-")
+	if len(segs) != 2 || len(segs[0]) == 0 || len(segs[1]) == 0 {
+		return nil, fmt.Errorf("invalid resource property name: %s", name)
+	}
+	cn, tn := segs[0], segs[1]
+
+	// Query the resource usage data.
+	var q resource.Quantity
+	var found bool
+	switch cn {
+	case totalCapacityName:
+		// The property concerns the total capacity of a resource.
+		q, found = cluster.Status.ResourceUsage.Capacity[corev1.ResourceName(tn)]
+	case allocatableCapacityName:
+		// The property concerns the allocatable capacity of a resource.
+		q, found = cluster.Status.ResourceUsage.Allocatable[corev1.ResourceName(tn)]
+	case availableCapacityName:
+		// The property concerns the available capacity of a resource.
+		q, found = cluster.Status.ResourceUsage.Available[corev1.ResourceName(tn)]
+	default:
+		// The property concerns a capacity type that cannot be recognized.
+		return nil, fmt.Errorf("invalid capacity type %s in resource property name %s", cn, name)
+	}
+
+	if !found {
+		// The property concerns a resource that is not present in the resource usage data.
+		//
+		// It cound be that the resource is not available in the cluster; consequently Fleet
+		// does not consider this as an error.
+		return nil, nil
+	}
+	return &q, nil
 }
 
-// AffinityTerms is a "processed" representation of []ClusterSelectorTerms.
-// The terms are `ORed`.
-type AffinityTerms []affinityTerm
+// retrievePropertyValueFrom retrieves a property value, resource or non-resource,
+// from a member cluster.
+//
+// Note that it will return nil if the property is not available for the cluster;
+// the zero value of resource.Quantity, i.e., resource.Quantity{}, is a valid
+// quantity.
+func retrievePropertyValueFrom(cluster *clusterv1beta1.MemberCluster, name string) (*resource.Quantity, error) {
+	// Check if the expression concerns a resource property.
+	var q *resource.Quantity
+	var err error
+	if strings.Contains(name, resourcePropertyNamePrefix) {
```

Comment:
if this is the prefix, why not check if the string starts with instead of contains? We didn't even check the error in the next line.

Reply:
Done.